 **Modelo de Pre Aprobación**

Caso de estudio: Se requiere implementar un modelo de pre aprobación de créditos a partir de las variables socio económicas de un solicitante de crédito.

* Edad: Indica el número de años de un solicitante de crédito
* Ingresos: Indica todos los ingresos que recibe una persona
* Egresos: Indica todos los hastos mensuales de una persona
*Monto (EAD) : Es el monto que una persona está solicitando a través del crédito

La variable de referencia será la variable PreAprobación. Esta variable indica la pre aprobacion o pre negación de un crédito.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

2. Cargamos Datos de trabajo

In [ ]:
nxl = '/content/sample_data/6. SolicitantesCrédito(USD).xlsx'
XDB = pd.read_excel(nxl, sheet_name=0)

#Sekeccionamos las variables de trabajo
XDB=XDB[["Edad", "Ingresos", "Egresos","Monto (EAD)" , "PreApr"]]
XDB=XDB.dropna(subset= ["PreApr"])
display(XDB.head())

#Separar variables de entrada - variables de salida
XD=XDB[["Edad","Ingresos","Egresos","Monto (EAD)"]] #Entrada
yd=XDB["PreApr"] #Variable de salida


Edad    Ingresos      Egresos  Monto (EAD)  PreApr
0    38  1356.14400  1685.622667   625.562230     1.0
1    51   286.01600   364.462000   140.031984     1.0
2    37   517.46325   629.208889   284.564492     0.0
3    29   473.27000   551.616889   309.647588     1.0
4    42   750.09175   806.715778   500.663578     1.0

3. Se implementa el modelo de clasificación

In [ ]:
mnb=GaussianNB() #Coco de modelo
mnb.fit(XD,yd) #Entrenamos el modelo

#Datos por categoría
ndat=mnb.class_count_
print("El número de datos por categoría es: \n",ndat)
cat=mnb.classes_
print("Las categorías son: \n",cat)

#Determinamos la propiedad estadistica de las variables
u=mnb.theta_
print("La mededida de las variables es : \n:",u)
sigma=np.sqrt(mnb.var_)
print("La desviación de las variables es: \n", sigma)
print("Los límites superiores de las variables son \n" , u+sigma)
print("Los limites inferiores de las variables son: \n", u-sigma)

El número de datos por categoría es: 
 [2959. 2883.]
Las categorías son: 
 [0. 1.]
La mededida de las variables es : 
: [[  40.98411626  431.42680644  401.54995644  289.41445941]
 [  41.10752688 1029.98947251  900.19312144  781.2021157 ]]
La desviación de las variables es: 
 [[ 12.21303268 325.79791497 342.77286087 235.95783541]
 [ 12.25106493 472.43342807 525.43337837 469.90038088]]
Los límites superiores de las variables son 
 [[  53.19714893  757.22472141  744.32281731  525.37229482]
 [  53.35859181 1502.42290058 1425.62649981 1251.10249658]]
Los limites inferiores de las variables son: 
 [[ 28.77108358 105.62889147  58.77709557  53.456624  ]
 [ 28.85646196 557.55604444 374.75974307 311.30173482]]


Analisis de resultados: De acuerdo con la base de datos se puede identificar un total de 5842 datos, los cuales estan divididos en 2959 en pre negados y 2883 en pre aprobados se puede observar igualmente que la categoria

4.Se procede con la evaluación de desempeño del modelo (Frente al pronóstico de la decisión)

In [ ]:
ydp=mnb.predict(XB)
cm=confusion_matrix(yb,ydp)
print("La matriz de confusión es: \n",cm)

5.Métricas de desempeño

In [ ]:
cm=confusion_matrix(yb,ydp)
print("La matriz de confusión es: \n",cm)
VN=cm[0,0];FP=cm[0,1]
FN=cm[1,0];VP=cm[1,1]
#Tasa de error

TD=Len(XB) #Total datos

#Exactitud: Muestra el comportamiento general del modelo
Ex=(VN+VP)/TP,Print("La exactitud del modelo es: ",Ex)
#Tasa de error: Porcentaje de equivocación
Ter=(FN+FP)/TD; print("La tasa de error del modelo es: ",Ter)
#Sensibilidad:Como se comporta el modelo frente a los preAprobados
Sen=VP/(VP+FN); print("La sensibilidad del modelo es: ",Sen)
#Especificidad: Como se comporta el modelo frente a los preNegados
Esp=VN/(VN+FP); print("La especificidad del modelo es: ",Esp)
#Precisión: Capacidad del modelo para predecir VP
pre=VP/(VP+FP); print("La precisión del modelo es: ",pre)
#Precision Negativa: Capacidad del modelo para predecir VN
pre_neg=VN/(VN+FN); print("La precisión negativa del modelo es: ",pre_neg)

6

7. Creamos un informe con la/los que no poseen pre aprobación

In [ ]:
nxl="/content/6. SolicitantesCrédito(USD).xlsx"
XDB=pd.read_excel(nxl,sheet_name=0)

#Seleccionamos las variables de trabajo
XDB=XDB[["Edad","Ingresos","Egresos","Monto (EAD)","PreApr"]]

filas_sp=np.where(np.isnan(XDB["PreApr"]))[0]
XDBsp=XDB.iloc[filas_sp,]
XDBsp=XDBsp[["Edad","Ingresos","Egresos","Monto(EAD)"]]

#Procedemos con el pronóstico
ydpsp=mnb.predict(XDBsp)

#Pegamos el pronostico
TAsq=np.column_stack((XDBsp,ydpsp))
df=pd.DataFrame(TAsq)
df.columns=["Solicitante","Edad","Ingresos","Egresos","Monto(EAD)","PreApr"]
display(df)


FileNotFoundError: [Errno 2] No such file or directory: '/content/6. SolicitantesCrédito(USD).xlsx'